postgis/lwgeom_geos.c

```diff
@@ -846,51 +846,105 @@ Datum boundary(PG_FUNCTION_ARGS)
 		PG_RETURN_NULL();
 	}
 
-	/* GEOS doesn't do triangle type, so we special case that here */
-	if (lwgeom->type == TRIANGLETYPE)
+	switch(lwgeom->type)
 	{
-		lwgeom->type = LINETYPE;
-		result = geometry_serialize(lwgeom);
-		lwgeom_free(lwgeom);
-		PG_RETURN_POINTER(result);
-	}
-
-	initGEOS(lwpgnotice, lwgeom_geos_error);
-
-	g1 = LWGEOM2GEOS(lwgeom, 0);
-	lwgeom_free(lwgeom);
-
-	if (!g1)
-		HANDLE_GEOS_ERROR("First argument geometry could not be converted to GEOS");
+		case POINTTYPE:
+		{
+			LWPOINT *lwpoint = (LWPOINT *)lwgeom;
+			lwpoint->point->npoints = 0;
+			break;
+		}
+		case MULTIPOINTTYPE:
+		{
+			LWMPOINT *lwmpoint = (LWMPOINT *)lwgeom;
+			lwmpoint->ngeoms = 0;
+			break;
+		}
+		case LINETYPE:
+		{
+			POINTARRAY *pa = ((LWLINE *)lwgeom)->points;
+			ptarray_copy_point(pa, pa->npoints - 1, 1);
+			pa->npoints = 2;
 
-	g3 = GEOSBoundary(g1);
+			LWGEOM *lwmpoint = (LWGEOM *)lwmpoint_from_lwgeom(lwgeom);
+			lwgeom_free(lwgeom);
+			lwgeom = lwmpoint;
 
-	if (!g3)
-	{
-		GEOSGeom_destroy(g1);
-		HANDLE_GEOS_ERROR("GEOSBoundary");
-	}
+			break;
+		}
+		case MULTILINETYPE:
+		{
+			LWGEOM *lwmpoint = (LWGEOM *)lwmpoint_from_lwgeom(lwgeom);
+			// it is not going to work
+			lwgeom_remove_repeated_points_in_place(lwmpoint, 0.0001);
```

Comment:
points that occur even number of times should be left and points that occur odd number of times should be removed. am I right?

Reply:
I believe it was about endpoints only, not any points.

Reply:
oh, yes, you are right. so only endpoints that occur odd number of times should be included in result.